### PR TITLE
Unify text generation interface and test pipeline to use same function for Cloud AI 100 execution

### PR DESCRIPTION
**Pull request by quic-mamta**
1. Unify text generation interface and test pipeline to use same function for Cloud AI 100 execution
2. Unify generate_inputs and export_hf_to_cloud_ai_10
3. Split check_num_prompts_and_batch_size function to get input prompts seperately
4. Update compile API to take custom_io_file_path as optional input from user
5. Update HL API tests
6. Support running in loop for BS>1 in cloud_ai_100_exec_kv
7. Fixed mismatch in tokens during export
8. Code clean up in tests/utils.py
9. Add Phi3 in tests(config.json) and use 2 layer models for testing

### CODE REVIEW

File: QEfficient/generation/text_generation_inference.py [hunk 1]
```diff
@@ -144,15 +144,17 @@ def cloud_ai_100_exec_kv_helper(
     enable_debug_logs: bool = False,
     stream: bool = True,
     write_io_dir: Optional[str] = None,
+    session=None,
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:147]: why is this being added?

Comment (ochougul) [reply QEfficient/generation/text_generation_inference.py:147]: Can't we change utils to pass `qpc_path` instead of `session`?

File: QEfficient/utils/run_utils.py [hunk 2]
```diff
@@ -162,28 +161,24 @@ def run_kv_model_on_ort(self, model_path, n_layer, padding_shape):
...
-
-        for i in range(1, self.gen_len):
-            generated_ids.append(outputs["logits"].argmax(-1).reshape(-1, 1))
-            inputs = self.input_handler.update_cloud_ai_100_inputs(i, inputs, outputs)
-            session.skip_buffers([x for x in session.input_names + session.output_names if x.startswith("past_")])
-            outputs = session.run(inputs)
-
-        generated_ids.append(outputs["logits"].argmax(-1).reshape(-1, 1))
-        generated_ids = np.concatenate(generated_ids, axis=1)
+        from QEfficient.generation.text_generation_inference import cloud_ai_100_exec_kv_helper
```
Comment (ochougul) [inline QEfficient/utils/run_utils.py:172]: move to top

File: QEfficient/generation/text_generation_inference.py (hunk 1 above)
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:147]: `input_len` variable is not being used anywhere.

Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:147]: If we pass qpc_path then we need to create session inside run_kv_model_on_cloud_ai_100 function in run_util.py, for which we need to import QAICInferenceSession, When running tests it will fail during this import if machine does not have aic card.

File: QEfficient/generation/text_generation_inference.py [hunk 3]
```diff
@@ -237,7 +240,7 @@ def cloud_ai_100_exec_kv_helper(
     print()
 
     latency_stats = (generated_texts, prefill_perf, decode_perf, total_perf, total_time)
-    return latency_stats
+    return latency_stats, generated_ids
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:243]: Create a `dataclass` that holds all the info about `Cloud AI 100 execution` and return it from here.
Also, make `cloud_ai_100_exec_kv` return the same `dataclass` object.

```
Class CloudAI100ExecInfo
    generated_texts: str
    generated_ids: List[int]
    prefil_perf: float
    decode_perf: float
    ...
```

File: QEfficient/generation/text_generation_inference.py (hunk 1 above)
Comment (ochougul) [reply QEfficient/generation/text_generation_inference.py:147]: Don't we skip the ai 100 execution test anyway when device is not present?

File: QEfficient/utils/run_utils.py (hunk 2 above)
Comment (quic-mamta) [reply QEfficient/utils/run_utils.py]: Updated

File: QEfficient/generation/text_generation_inference.py (hunk 1 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:158]: Updated

File: QEfficient/generation/text_generation_inference.py (hunk 3 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py]: Updated

File: QEfficient/generation/text_generation_inference.py (hunk 1 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py]: If machine does not have aic card, it fails in get_available_device_id in device_utils.py, sorry for the confusion, changed it to take qpc path instead of session.

File: QEfficient/cloud/compile.py [hunk 4]
```diff
@@ -44,6 +44,12 @@
         type=int,
         help="num cores to compile the model on",
     )
+    parser.add_argument(
```
Comment (ochougul) [inline QEfficient/cloud/compile.py:47]: This is being added for backward compatibility with the onnx files generated by `RunKV` script, where `custom_io` file doesn't exist where `qeff` expects.

File: QEfficient/generation/text_generation_inference.py [hunk 5]
```diff
@@ -18,6 +19,17 @@
 from QEfficient.utils import padding_check_and_fix
 from QEfficient.utils.logging_utils import logger
 
+
+@dataclass
+class CloudAI100ExecInfo:
+    generated_texts: List[str] = None
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:25]: Why are you assigning None, can there be a case where one/all will not be passed while initializing this class?

File: QEfficient/generation/text_generation_inference.py [hunk 6]
```diff
@@ -269,68 +289,66 @@ def print_latency_stats_kv(
...
-
     print_latency_stats_kv(
         prompt,
-        generated_texts,
-        batch_size,
-        prefill_time,
-        decode_perf,
-        total_perf,
-        total_time,
+        generated_texts=execinfo.generated_texts,
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:346]: pass `prompt, execinfo, automation` handle printing the attributes inside `print` function.

File: QEfficient/generation/text_generation_inference.py [hunk 7]
```diff
@@ -18,6 +19,17 @@
 from QEfficient.utils import padding_check_and_fix
 from QEfficient.utils.logging_utils import logger
 
+
+@dataclass
+class CloudAI100ExecInfo:
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:24]: Please add docstring.

File: QEfficient/generation/text_generation_inference.py (hunk 5 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:25]: It is returned in cloud_ai_100_exec_kv and cloud_ai_100_exec_kv_helper. in both places no params will be None, updating this.

File: QEfficient/generation/text_generation_inference.py (hunk 7 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py:24]: Updating this.

File: QEfficient/generation/text_generation_inference.py (hunk 6 above)
Comment (quic-mamta) [reply QEfficient/generation/text_generation_inference.py]: updated

**Review by ochougul — COMMENTED** [review]
update readme API usage

**Comment by ochougul** [comment]
Please rebase

File: QEfficient/generation/text_generation_inference.py [hunk 8]
```diff
@@ -235,111 +266,112 @@ def cloud_ai_100_exec_kv_helper(
...
-            total_time.append(latency_stats[4])
-
-        prefill_time = np.average(prefill_time)
-        decode_perf = np.average(decode_perf)
-        total_perf = np.average(total_perf)
-        total_time = np.average(total_time)
-
-    else:
-        latency_stats = cloud_ai_100_exec_kv_helper(
+    batch_size, ctx_len = get_compilation_dims(qpc_path)
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:329]: Add 
```
if os.path.exists(specialization_file_path):
    open(file)
else:
   raise FileNotFoundError(f"expected specializations.json file at path, {qpc_base_path}")

```
inside `get_compilation_dims` functions.

File: QEfficient/compile/compile_helper.py [hunk 9]
```diff
@@ -139,11 +141,12 @@ def compile(
...
         custom_io_file_name = "custom_io_fp16.yaml"
 
-    custom_io_file_path = os.path.join(os.path.dirname(onnx_path), custom_io_file_name)
+    if custom_io_file_path is None:
+        custom_io_file_path = os.path.join(os.path.dirname(onnx_path), custom_io_file_name)
 
     if not os.path.isfile(custom_io_file_path):
         raise FileNotFoundError(
-            f"file {custom_io_file_path} needs to exist in the same directory as onnx model files. Please rerun infer/export Api"
+            f"Custom IO file {custom_io_file_name} is neither present at given path nor exists in the onnx model files directory. Please pass the correct file path or rerun infer/export Api"
```
Comment (ochougul) [inline QEfficient/compile/compile_helper.py:149]: f"Custom IO file {custom_io_file_name} is not present at the expected path {custom_io_file_path} nor exists in the onnx model files directory. Please pass the correct file path or rerun infer/export Api"

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 10]
```diff
@@ -193,54 +194,17 @@ def export_kvstyle_transformed_model_to_onnx(
 
     # Check and fix tokenizer viability
     padding_check_and_fix(tokenizer)
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:196]: remove this, anyway being done in InputHandler.

File: QEfficient/utils/_utils.py [hunk 11]
```diff
@@ -188,3 +188,40 @@ def padding_check_and_fix(tokenizer: Union[PreTrainedTokenizer, PreTrainedTokeni
             tokenizer.pad_token_id = tokenizer.eos_token_id
         else:
             tokenizer.pad_token_id = tokenizer.vocab_size - 1
+
+
+def get_config(config):
```
Comment (ochougul) [inline QEfficient/utils/_utils.py:193]: change this function to return `padding_shape`, and rename it as `get_padding_shape_from_config`

Comment (quic-mamta) [reply QEfficient/utils/_utils.py:193]: this function returns number of layers along with other parameters needed for padding shape.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 12]
```diff
@@ -251,16 +215,10 @@ def export_kvstyle_transformed_model_to_onnx(
...
-    # todo: vbaddi, fix it to verify on Cloud AI 100.
-    for i in range(1):
-        pt_outputs = transformed_model(**inputs)
-        inputs["input_ids"] = pt_outputs.logits.detach().argmax(2)
-        inputs["position_ids"] += 1
+    inputs = input_handler.update_pytorch_inputs(inputs, pt_outputs)
+
     # To avoid issues in onnx export
-    inputs["position_ids"] = torch.full((batch_size, 1), seq_len - 1)
+    inputs["position_ids"] = torch.full((1, 1), seq_len - 1)
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:221]: Let's discuss this change.

File: QEfficient/exporter/export_utils.py [hunk 13]
```diff
@@ -319,12 +319,12 @@ def run_model_on_ort(
         past_value_mean = past_key_sum / num
         print(f"past_keys (mean) \t\t {past_key_mean}")
         print(f"past_value (mean) \t\t {past_value_mean}")
-        print("\n=============================================================\n")
+        print("\n=====================================================================\n")
```
Comment (ochougul) [inline QEfficient/exporter/export_utils.py:322]: Let's remove print statements and use `logging.info`

File: QEfficient/generation/text_generation_inference.py [hunk 14]
```diff
@@ -18,6 +19,27 @@
...
+    :total_time: float
+    """
+
+    generated_texts: List[str]
+    generated_ids: np.ndarray
+    prefill_time: float
+    decode_perf: float
+    total_perf: float
+    total_time: float
+
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:41]: add `__repr__` method, which can be used for printing the stats of execution just by printing or logging the object of this class.

File: QEfficient/generation/text_generation_inference.py [hunk 15]
```diff
@@ -235,111 +266,112 @@ def cloud_ai_100_exec_kv_helper(
...
+def print_latency_stats_kv(prompt, batch_size, execinfo, automation: bool = False):
     if automation:
         print()
         print("input=", prompt)
-        print("output=", generated_texts)
-        print("Prefill time a.k.a TTFT is=", round(prefill_time, 2))
-        print("Decode token/sec is=", round(decode_perf * batch_size, 2))
-        print("Total token/sec is=", round(total_perf * batch_size, 2))
-        print("Total (E2E) inference time is=", round(total_time, 2))
+        print("output=", execinfo.generated_texts)
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:296]: Just do `print(execinfo)` here and move this code in `__repr__` method of the dataclass

File: QEfficient/generation/text_generation_inference.py [hunk 16]
```diff
@@ -118,15 +140,25 @@ def check_batch_size_and_num_prompts(prompt, prompts_txt_file_path, batch_size)
...
-
-    num_prompts = len(prompt)
-    if batch_size > 1:
-        assert (
-            batch_size == num_prompts
-        ), f"Mismatch between number of prompts {num_prompts} and batch size {batch_size}; please pass correct input argument"
     return prompt
 
 
+def check_batch_size_and_num_prompts(prompt: List[str], batch_size: int):
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:146]: rename to `fix_prompts_and_get_num_iters`.
You don't need to return prompt, as list are passed by reference.
return only `num_iterations`

File: QEfficient/generation/text_generation_inference.py [hunk 17]
```diff
@@ -185,10 +213,13 @@ def cloud_ai_100_exec_kv_helper(
             "Make sure this model supports sliding window, such as Mistral"
         )
     assert generation_len > 0, "generation length should be greater than zero"
-    generated_ids = np.full((batch_size, generation_len + 1), tokenizer.pad_token_id)
+    generated_ids = np.full((batch_size, generation_len), tokenizer.pad_token_id)
     if stream:
         streamer = transformers.TextStreamer(tokenizer)
-        streamer.on_finalized_text(prompt[0] + " ")
+        print()
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:219]: are these print statement required?

File: QEfficient/utils/_utils.py (hunk 11 above)
Comment (ochougul) [reply QEfficient/utils/_utils.py:193]: Move this call inside InputHandler initializer as discussed.

File: README.md [hunk 18]
```diff
@@ -183,12 +183,12 @@ python -m QEfficient.cloud.infer --model_name gpt2 --batch_size 1 --prompt_len 3
...
+|  QEfficient.transform    |   [click here](#1-model-download-and-transform)         |  <li>model : $\color{green} {Mandatory}$</li><li>form_factor : Optional [Default="cloud"]</li>  | 
+| QEfficient.export |     [click here](#2-onnx-export-of-transformed-model)       |   <li>model_name : $\color{green} {Mandatory}$</li><li>model_kv : Optional </li><li>local_model_dir : Optional [Path to custom model weights and config file]</li> <li>tokenizer : Optional</li> <li>cache_dir : Optional ["cache_dir" in current working directory]</li> <li>onnx_dir_path : Optional </li><li>hf_token : Optional</li><li>seq_length : Optional [Default=128]</li><li>kv : Optional [Default=True]</li><li>form_factor : Optional [Default="cloud"]</li>
```
Comment (ochougul) [inline README.md:187]: cache_dir : Optional [Path to HuggingFace cache dir if changed from default value]

File: README.md [hunk 19]
```diff
@@ -114,8 +114,8 @@ In summary:
...
+| QEfficient.cloud.infer           |   [click here](#1-use-qefficientcloudinfer)         |  <li>model_name : $\color{green} {Mandatory}$</li> <li>num_cores : $\color{green} {Mandatory}$</li> <li>device_group : $\color{green} {Mandatory}$</li> <li>**prompt : Optional</li><li>**prompts_txt_file_path : Optional</li> <li>aic_enable_depth_first : Optional </li> <li>mos : Optional [Default=-1]</li> <li>batch_size : Optional [Default=1]</li> <li>prompt_len : Optional [Default=32]</li> <li>ctx_len : Optional [Default=128]</li> <li>generation_len : Optional [Default=None]</li> <li>mxfp6 : Optional </li> <li>mxint8 : Optional </li> <li>local_model_dir : Optional [Path to custom model weights and config file]</li> <li>cache_dir : Optional ["cache_dir" in current working directory]</li> <li>hf_token : Optional </li> <li>verbose : Optional</li>|
```
Comment (ochougul) [inline README.md:117]: Path to the directory used for saving HuggingFace cache, Default is "efficient-transformers/cache_dir".

File: tests/config.json [hunk 20]
```diff
@@ -16,6 +16,10 @@
       "model_name": "microsoft/phi-2",
       "model_class": "PhiForCausalLM"
     },
+    {
+      "model_name": "microsoft/Phi-3-mini-4k-instruct",
+      "model_class": "Phi3ForCausalLM"
```
Comment (ochougul) [inline tests/config.json:21]: Add extra parameter named `num_layers`, in case it's provided it will be used in tests, if not provided default value will be 1.

File: tests/utils.py [hunk 21]
```diff
@@ -137,7 +116,8 @@ def set_up(model_config, device_group=[0]):
     except Exception as e:
         print(f"Pytorch HuggingFace Pytorch Model run failed due to : {e}")
 
-    model_kv = transform_pt_model_with_qeff(model_hf)
+    model_kv = transform_lm(model_hf)
+    model_kv.eval()
```
Comment (ochougul) [inline tests/utils.py:120]: remove if not required.

File: tests/config.json (hunk 20 above)
Comment (quic-mamta) [reply tests/config.json:21]: to be done in PR #79

File: QEfficient/exporter/export_utils.py (hunk 13 above)
Comment (quic-mamta) [reply QEfficient/exporter/export_utils.py:322]: todo @ochougul

File: QEfficient/generation/text_generation_inference.py [hunk 22]
```diff
@@ -235,111 +272,102 @@ def cloud_ai_100_exec_kv_helper(
...
+        batch_size=batch_size,
+        generated_texts=generated_texts,
+        generated_ids=generated_ids,
+        prefill_time=prefill_time,
+        decode_perf=decode_perf,
+        total_perf=total_perf,
+        total_time=total_time,
+    )
     print_latency_stats_kv(
         prompt,
```
Comment (carlstreeter-quic) [inline QEfficient/generation/text_generation_inference.py:369]: This is (still) printing to stdout, rather than just returning.  Could we just return at this layer, so we have a clean API to call programmatically?

Comment (ochougul) [reply QEfficient/generation/text_generation_inference.py:369]: Hi @carlstreeter-quic , we want to support logger for printing/saving such info in log fiiles.
So, user can setup logger as they want and handle this info, it won't `print` by default.
We plan to do this as follow up to this PR.
This PR has already grown a lot.
